doc/man7/provider-keymgmt.pod

```diff
@@ -83,133 +64,143 @@ All these "functions" have a corresponding function type definition
- OP_keymgmt_get_key_params       OSSL_FUNC_KEYMGMT_GET_KEY_PARAMS
- OP_keymgmt_gettable_key_params  OSSL_FUNC_KEYMGMT_GETTABLE_KEY_PARAMS
+ OP_keymgmt_new                  OSSL_FUNC_KEYMGMT_NEW
+ OP_keymgmt_free                 OSSL_FUNC_KEYMGMT_FREE
+
+ OP_keymgmt_get_params           OSSL_FUNC_KEYMGMT_GET_PARAMS
+ OP_keymgmt_gettable_params      OSSL_FUNC_KEYMGMT_GETTABLE_PARAMS
+
+ OP_keymgmt_has_parameters       OSSL_FUNC_KEYMGMT_HAS_PARAMETERS
+ OP_keymgmt_has_public_key       OSSL_FUNC_KEYMGMT_HAS_PUBLIC_KEY
+ OP_keymgmt_has_private_key      OSSL_FUNC_KEYMGMT_HAS_PRIVATE_KEY
 
  OP_keymgmt_query_operation_name OSSL_FUNC_KEYMGMT_QUERY_OPERATION_NAME
 
- OP_keymgmt_validate_domparams   OSSL_FUNC_KEYMGMT_VALIDATE_DOMPARAMS
+ OP_keymgmt_import               OSSL_FUNC_KEYMGMT_IMPORT
+ OP_keymgmt_import_types         OSSL_FUNC_KEYMGMT_IMPORT_TYPES
+ OP_keymgmt_export               OSSL_FUNC_KEYMGMT_EXPORT
+ OP_keymgmt_export_types         OSSL_FUNC_KEYMGMT_EXPORT_TYPES
+
+ OP_keymgmt_validate_parameters  OSSL_FUNC_KEYMGMT_VALIDATE_PARAMETERS
  OP_keymgmt_validate_public      OSSL_FUNC_KEYMGMT_VALIDATE_PUBLIC
  OP_keymgmt_validate_private     OSSL_FUNC_KEYMGMT_VALIDATE_PRIVATE
  OP_keymgmt_validate_pairwise    OSSL_FUNC_KEYMGMT_VALIDATE_PAIRWISE
 
-=head2 Domain Parameter Functions
+=head2 Key Objects
 
-OP_keymgmt_importdomparams() should create a provider side structure
-for domain parameters, with values taken from the passed B<OSSL_PARAM>
-array I<params>.
+A key object is a collection of data for an asymmetric key, and is
+represented as I<keydata> in this manual.
 
-OP_keymgmt_gendomparams() should generate domain parameters and create
-a provider side structure for them.
-Values of the passed B<OSSL_PARAM> array I<params> should be used as
-input for parameter generation.
+The exact contents of a key object are defined by the provider, and it
+is assumed that different operations in one and the same provider use
+the exact same structure to represent this collection of data, so that
+for example, a key object that has been created using the KEYMGMT
+interface that we document here can be passed as is to other provider
+operations, such as OP_signature_sign_init() (see
+L<provider-signature(7)>).
 
-OP_keymgmt_freedomparams() should free the passed provider side domain
-parameter structure I<domparams>.
+With some of the KEYMGMT functions, it's possible to select a specific
+subset of data to handle, governed by the bits in a I<selection>
+indicator.  The bits are:
 
-OP_keymgmt_exportdomparams() should extract values from the passed
-provider side domain parameter structure I<domparams> into the passed
-B<OSSL_PARAM> I<params>.
-Only the values specified in I<params> should be extracted.
+=over 4
 
-OP_keymgmt_importdomparam_types() should return a constant array of
-descriptor B<OSSL_PARAM>, for parameters that OP_keymgmt_importdomparams()
-can handle.
+=item B<OSSL_KEYMGMT_B_PRIVATE_KEY>
```

Comment:
Why `_B_*`? Wouldn't it be nicer to have uniform (if somewhat longer) names: 
```
OSSL_KEYMGMT_SELECT_PRIVATE_KEY
OSSL_KEYMGMT_SELECT_PUBLIC_KEY
OSSL_KEYMGMT_SELECT_KEY
OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS
OSSL_KEYMGMT_SELECT_OTHER_PARAMETERS
OSSL_KEYMGMT_SELECT_PARAMETERS
OSSL_KEYMGMT_SELECT_ALL
```

rather than 
```
OSSL_KEYMGMT_B_PRIVATE_KEY
OSSL_KEYMGMT_B_PUBLIC_KEY
OSSL_KEYMGMT_SELECT_KEY
OSSL_KEYMGMT_B_DOMAIN_PARAMETERS
OSSL_KEYMGMT_B_OTHER_PARAMETERS
OSSL_KEYMGMT_SELECT_PARAMETERS
OSSL_KEYMGMT_SELECT_ALL
```

Reply:
I wanted a way to denote that some of them are distinct bits, while the other are collections of bits.  For example, `OSSL_KEYMGMT_SELECT_KEY` = `OSSL_KEYMGMT_B_PRIVATE_KEY | OSSL_KEYMGMT_B_PUBLIC_KEY`

Reply:
We have used ```_FLAG_``` where that is important.  I would do ```OSSL_FLAG_PRIVATE_KEY``` I think the "keymgmt" context is obvious :)

Reply:
I got that, but I still find the "`_B_`" confusing. Also, in some way, `{domain|other} parameterS` are already collections, no? (I get that they might still get returned in a single object, but the collection distinction is thin, IMO).

Reply:
None of the bits are meant to denote one single item.  With RSA for example, the public key is *n* + *e* and the private key is *d* + *e* at a minimum, right?  So every bit represents a distinct category of stuff, and the `_SELECT_` values denote a set of categories.  Does saying it like this help?

Re `_B_` vs `_FLAG_`, I have no problems changing to the latter if that is more pleasing.  However, I will *not* drop `_KEYMGMT_`, that's just *asking* for future clashes.

Reply:
Side note: regarding "domain" and "other" parameters, it comes from the dispute on how to view key parameters, where some are considered domain parameters (the FFC parameters, for example), while others aren't (PSS and OAEP parameters have been mentioned as such), so this distinction is really catering for those who do want to make a distinction...  as for "other", I simply couldn't figure out a better name.

Reply:
I strongly doubt it's asking for future clashes, but not my decision.

Reply:
I would just leave all the names consistent.. They are all flags regardless of whether they are a collection of flags or not. Even doing this would be better..

OSSL_KEYMGMT_FLAG_DOMAIN_PARAMETERS
OSSL_KEYMGMT_FLAG_OTHER_PARAMETERS
OSSL_KEYMGMT_FLAG_PUBLIC_KEY
OSSL_KEYMGMT_FLAG_PRIVATE_KEY

OSSL_KEYMGMT_FLAGS_KEY
OSSL_KEYMGMT_FLAGS_PARAMETERS
OSSL_KEYMGMT_FLAGS_ALL

Reply:
Actually, I'm not sure I like where this is going. The *selection* is fairly special purpose, while "FLAG" doesn't quite reflect that.

OSSL_KEYMGMT_SEL_DOMAIN_PARAMETERS
OSSL_KEYMGMT_SEL_OTHER_PARAMETERS
OSSL_KEYMGMT_SEL_PUBLIC_KEY
OSSL_KEYMGMT_SEL_PRIVATE_KEY

OSSL_KEYMGMT_SEL_ALL_KEYS
OSSL_KEYMGMT_SEL_ALL_PARAMETERS
OSSL_KEYMGMT_SEL_ALL

Something like that? SEL or SELECT?

Reply:
+1 SELECT

Reply:
I actually like what @h-vetinari suggested above. ALL_KEYS could be _KEYPAIR..
Either way I dont want you getting stuck on names, when there are lots of other things to do.

Reply:
Hahaha, so after whining and moaning for a bit, I'll end up doing almost exactly what @h-vetinari suggests... I feel a bit embarrassed having whined 😉

Reply:
Okie, done!

Reply:
(or almost...)